tests/testthat/test_loo_moment_matching.R

```diff
@@ -0,0 +1,303 @@
+library(loo)
+library(rstan)
+options(mc.cores = 1)
+
+
+context("moment matching")
+set.seed(123)
+S <- 40000
+
+# helper functions for sampling from the posterior distribution
+rinvchisq <- function(n, df, scale = 1/df)
+{
+  if ((length(scale) != 1) & (length(scale) != n))
+    stop("scale should be a scalar or a vector of the same length as x")
+  if (df <= 0)
+    stop("df must be greater than zero")
+  if (any(scale <= 0))
+    stop("scale must be greater than zero")
+  return((df*scale)/rchisq(n, df = df))
+}
+
+dinvchisq <- function(x, df, scale=1/df, log = FALSE)
+{
+  if (df <= 0)
+    stop("df must be greater than zero")
+  if (scale <= 0)
+    stop("scale must be greater than zero")
+  nu <- df/2
+  if (log)
+    return(ifelse(x > 0, nu*log(nu) - log(gamma(nu)) + nu*log(scale) -
+                    (nu + 1)*log(x) - (nu*scale/x), NA))
+  else
+    return(ifelse(x > 0,
+                  (((nu)^(nu))/gamma(nu)) * (scale^nu) *
+                    (x^(-(nu + 1))) * exp(-nu*scale/x), NA))
+}
+
+
+# generate toy data
+# normally distributed data with known variance
+data_sd <- 1.1
+data_mean <- 1.3
+n <- as.integer(30)
+y <- rnorm(n = n, mean = data_mean, sd = data_sd)
+y_tilde <- 11
+y[1] <- y_tilde
+
+ymean <- mean(y)
+s2 <- sum((y - ymean)^2)/(n - 1)
+
+# draws from the posterior distribution when including all observations
+draws_full_posterior_sigma2 <- rinvchisq(S, n - 1, s2)
+draws_full_posterior_mu <- rnorm(S, ymean, sqrt(draws_full_posterior_sigma2/n))
+
+
+# create a dummy model object
+x <- list()
+x$data <- list()
+x$data$y <- y
+x$data$n <- n
+x$data$ymean <- ymean
+x$data$s2 <- s2
+
+x$draws <- data.frame(
+  mu = draws_full_posterior_mu,
+  sigma = sqrt(draws_full_posterior_sigma2)
+)
+
+
+
+
+
+
+
+
+# implement functions for moment matching loo
+
+
+# extract original posterior draws
+post_draws_test <- function(x) {
```

Comment:
all of these functions should contain a `...` argument even if not actually used as otherwise they will break if any kind of additional argument is passed to one of these functions.

Reply:
fixed